test/common/common/log_macros_test.cc

```diff
@@ -37,4 +37,48 @@ TEST(Logger, All) {
   // Misc logging with no facility.
   ENVOY_LOG_MISC(info, "fake message");
 }
+
+TEST(Logger, evaluateParams) {
+  uint32_t i = 1;
+
+  // set logger's level to low level.
+  // log message with higher severity and make sure that params were evaluated.
+  GET_MISC_LOGGER().set_level(spdlog::level::info);
+  ENVOY_LOG_MISC(warn, "test message '{}'", i++);
+  ASSERT_THAT(i, testing::Eq(2));
+}
+
+TEST(Logger, doNotEvaluateParams) {
+  uint32_t i = 1;
+
+  // set logger's logging level high and log a message with lower severity
+  // params should not be evaluated.
+  GET_MISC_LOGGER().set_level(spdlog::level::critical);
+  ENVOY_LOG_MISC(error, "test message '{}'", i++);
+  ASSERT_THAT(i, testing::Eq(1));
+}
+
+TEST(Logger, logAsStatement) {
+  // Just log as part of if ... statement
+
+  if (true)
+    ENVOY_LOG_MISC(critical, "test message 1");
+  else
+    ENVOY_LOG_MISC(critical, "test message 2");
+
+  // do the same with curly brackets
+  if (true) {
+    ENVOY_LOG_MISC(critical, "test message 3");
+  } else {
+    ENVOY_LOG_MISC(critical, "test message 4");
+  }
+}
+
+#ifdef NVLOG
```

Comment:
Does the CI ever run in a mode with `NVLOG` defined?  Put another way, will this test ever be executed by the CI or will it always be compiled out?

Reply:
CI probably never runs with NVLOG defined. I invoked the test manually by running 
bazel test --copt=-DNVLOG //test/common/common/log_macros_test

Reply:
Should we set one of the circleCI jobs to define NVLOG?  Maybe the "release" job?

Reply:
That is probably good idea. I was surprised when my first PR #2676 was reverted even though it passed CI checks. If "release" CI job tests release image, it should probably use "release" compile flags.

Reply:
Tracking here: https://github.com/envoyproxy/envoy/issues/2754

Reply:
Can we create a separate test for this and make bazel add that copt in by adding it to the BUILD file?  https://docs.bazel.build/versions/master/be/c-cpp.html#cc_test

Reply:
I think it is safer to compile everything with same flags.

Reply:
@ggreenway I don't think we should do it on the release job because that job is used for building the docker images. 

Not to throw a wrench in this, but I'm wondering with all the macros we have now and this change, if we should just kill NVLOG entirely. There is some perf benefit but now that the logs don't evaluate it's probably small. 

I think all that would be needed is to add some type of macro/guard for the few places where we use NVLOG to block an expensive iteration. Thoughts?

Reply:
Performance overhead is one function call and comparison between integers. Not much.
I just hope that trace and debug logs do not display sensitive information. With NVLOG they are just removed. If NVLOG is killed, all is required to display those logs is to lower severity level.

Reply:
I think at this point I would vote to just kill NVLOG and replace with some additional macros for the places in which we use it for larger guards, but no objection to merging this PR.

Reply:
+1

Reply:
@cpakulski are you interested in just getting rid of NVLOG as part of this PR?